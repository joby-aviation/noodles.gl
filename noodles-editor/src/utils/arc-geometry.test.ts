import { describe, expect, it } from 'vitest'
import {
  clamp,
  getArc,
  mix,
  mixspace,
  paraboloid,
  range,
  segmentRatios,
  smoothstep,
  tiltPoint,
} from './arc-geometry'

describe('mix function', () => {
  it('interpolates between two values', () => {
    expect(mix(0, 10, 0.5)).toBe(5)
    expect(mix(-10, 10, 0.25)).toBe(-5)
    expect(mix(5, 15, 1)).toBe(15)
    expect(mix(5, 15, 0)).toBe(5)
  })
})

describe('mixspace function', () => {
  it('returns an array of interpolated values', () => {
    expect(mixspace(0, 10, [0, 0.5, 1])).toEqual([0, 5, 10])
    expect(mixspace(-10, 10, [0.25, 0.75])).toEqual([-5, 5])
  })
})

describe('clamp function', () => {
  it('clamps values', () => {
    expect(clamp(5, 0, 10)).toBe(5)
    expect(clamp(-5, 0, 10)).toBe(0)
    expect(clamp(15, 0, 10)).toBe(10)
  })
})

describe('range function', () => {
  it('returns an array of sequential numbers', () => {
    expect(range(5)).toEqual([0, 1, 2, 3, 4])
    expect(range(0)).toEqual([])
  })
})

describe('smoothstep function', () => {
  it('returns a smooth transition', () => {
    expect(smoothstep(0, 1, 0)).toBe(0)
    expect(smoothstep(0, 1, 1)).toBe(1)
    expect(smoothstep(0, 1, 0.5)).toBeCloseTo(0.5, 2)
  })
})

describe('segmentRatios function', () => {
  it('returns correct segment ratios', () => {
    expect(segmentRatios(3, false)).toEqual([0, 0.5, 1])
  })
})

describe('paraboloid function', () => {
  it('returns correct altitude at given ratio', () => {
    expect(paraboloid(10, 5, 15, 0)).toBe(5)
    expect(paraboloid(10, 5, 15, 1)).toBe(15)
  })
})

describe('getArc function', () => {
  it('generates an arc with correct segment count', () => {
    const path = getArc({
      source: { lat: 0, lng: 0, alt: 0 },
      target: { lat: 10, lng: 10, alt: 10 },
      arcHeight: 100,
      segmentCount: 5,
    })

    expect(path.length).toBe(5)
  })

  it('generates an arc with correct apex height', () => {
    const path = getArc({
      source: { lat: 0, lng: 0, alt: 0 },
      target: { lat: 10, lng: 10, alt: 0 },
      arcHeight: 100,
      segmentCount: 5,
    })
    // TODO: This is actually expected to be equal to the arc height.. but it's 50%. This bug might now be a feature since many projects exist with it.
    expect(path[2][2]).toBeCloseTo(50, 1)
  })

  it('arcs work', () => {
    const path = getArc({
      source: { lat: 40.72145, lng: -73.9974, alt: 0 },
      target: { lat: 40.655, lng: -73.77585, alt: 0 },
      arcHeight: 10000,
      segmentCount: 11,
      tilt: 0,
    })
    expect(path.length).toBe(11)
    expect(path).toMatchSnapshot()
  })

  it('tilted arcs work', () => {
    const path = getArc({
      source: { lat: 40.72145, lng: -73.9974, alt: 0 },
      target: { lat: 40.655, lng: -73.77585, alt: 0 },
      arcHeight: 10000,
      segmentCount: 11,
      tilt: 90,
    })
    expect(path.length).toBe(11)
    expect(path).toMatchSnapshot()
  })
})

describe('tiltPoint function', () => {
  it('does not modify point if tilt is zero', () => {
    const point = [5, 5, 10]
    expect(tiltPoint(point, { lat: 0, lng: 0, alt: 0 }, { lat: 10, lng: 10, alt: 10 }, 0)).toEqual(
      point
    )
  })

  it('applies tilt', () => {
    const tilt1 = tiltPoint(
      [0, 0, 10],
      { lat: 0, lng: 0, alt: 0 },
      { lat: 10, lng: 10, alt: 10 },
      45
    )
    expect(tilt1[0]).toBeCloseTo(0, 3)
    expect(tilt1[1]).toBeCloseTo(0, 3)
    expect(tilt1[2]).toBeCloseTo(7.07, 1)
    const tilt2 = tiltPoint(
      [5, 5, 10],
      { lat: 0, lng: 0, alt: 0 },
      { lat: 10, lng: 10, alt: 10 },
      45
    )
    expect(tilt2[0]).toBeCloseTo(5, 3)
    expect(tilt2[1]).toBeCloseTo(5, 3)
    expect(tilt2[2]).toBeCloseTo(7.07, 1)
    const tilt3 = tiltPoint(
      [0, 0, 10],
      { lat: 0, lng: 0, alt: 0 },
      { lat: 10, lng: 10, alt: 10 },
      0.01
    )
    expect(tilt3[0]).toBeCloseTo(0, 3)
    expect(tilt3[1]).toBeCloseTo(0, 3)
    expect(tilt3[2]).toBeCloseTo(10, 3)
  })
})
