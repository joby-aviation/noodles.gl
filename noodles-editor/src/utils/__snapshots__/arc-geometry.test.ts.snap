// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getArc function > arcs work 1`] = `
[
  [
    -73.9974,
    40.72145,
    0,
  ],
  [
    -73.975245,
    40.714805,
    1649.7272501841032,
  ],
  [
    -73.95309,
    40.70816,
    3052.605444534226,
  ],
  [
    -73.930935,
    40.701515,
    4115.142767875739,
  ],
  [
    -73.90878000000001,
    40.69487,
    4775.939698111776,
  ],
  [
    -73.88662500000001,
    40.688225,
    5000,
  ],
  [
    -73.86447,
    40.68158,
    4775.939698111776,
  ],
  [
    -73.842315,
    40.674935,
    4115.14276787574,
  ],
  [
    -73.82016,
    40.66829,
    3052.6054445342243,
  ],
  [
    -73.798005,
    40.661645,
    1649.7272501841032,
  ],
  [
    -73.77585,
    40.655,
    0,
  ],
]
`;

exports[`getArc function > tilted arcs work 1`] = `
[
  [
    -73.9974,
    40.72145,
    0,
  ],
  [
    -73.96962164329857,
    40.7290159142383,
    1.0101665982020633e-13,
  ],
  [
    -73.94268574907245,
    40.73445544621441,
    1.8691817433543117e-13,
  ],
  [
    -73.91691068301331,
    40.73696324815505,
    2.519798209356702e-13,
  ],
  [
    -73.89250532493232,
    40.7360104184841,
    2.9244196321066814e-13,
  ],
  [
    -73.86958850753565,
    40.73129549615845,
    3.061616997868383e-13,
  ],
  [
    -73.84819857025437,
    40.722720418484094,
    2.9244196321066814e-13,
  ],
  [
    -73.8282962756151,
    40.71038324815505,
    2.5197982093567023e-13,
  ],
  [
    -73.80976197194592,
    40.69458544621441,
    1.8691817433543104e-13,
  ],
  [
    -73.7923861273568,
    40.675855914238305,
    1.0101665982020633e-13,
  ],
  [
    -73.77585,
    40.655,
    0,
  ],
]
`;
